c++/src/capnp/rpc-twoparty.c++

```diff
@@ -19,30 +19,53 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
-// Includes just for need SOL_SOCKET and SO_SNDBUF
-#if _WIN32
-#include <kj/win32-api-version.h>
-#endif
-
 #include "rpc-twoparty.h"
 #include "serialize-async.h"
 #include <kj/debug.h>
 #include <kj/io.h>
 
-#if _WIN32
-#include <winsock2.h>
-#include <mswsock.h>
-#include <kj/windows-sanity.h>
-#else
-#include <sys/socket.h>
-#endif
-
 namespace capnp {
 
-TwoPartyVatNetwork::TwoPartyVatNetwork(kj::AsyncIoStream& stream, rpc::twoparty::Side side,
-                                       ReaderOptions receiveOptions)
-    : stream(&stream), maxFdsPerMessage(0), side(side), peerVatId(4),
-      receiveOptions(receiveOptions), previousWrite(kj::READY_NOW) {
+TwoPartyVatNetwork::TwoPartyVatNetwork(
+    MessageStream& stream,
+    kj::Maybe<kj::Own<MessageStream>>&& ownStream,
+    uint maxFdsPerMessage,
+    rpc::twoparty::Side side,
+    ReaderOptions receiveOptions)
+
+    : stream(stream),
+      ownStream(kj::mv(ownStream)),
+      maxFdsPerMessage(maxFdsPerMessage),
+      side(side),
+      peerVatId(4),
+      receiveOptions(receiveOptions),
+      previousWrite(kj::READY_NOW) {
+  peerVatId.initRoot<rpc::twoparty::VatId>().setSide(
+      side == rpc::twoparty::Side::CLIENT ? rpc::twoparty::Side::SERVER
+                                          : rpc::twoparty::Side::CLIENT);
+
+  auto paf = kj::newPromiseAndFulfiller<void>();
+  disconnectPromise = paf.promise.fork();
+  disconnectFulfiller.fulfiller = kj::mv(paf.fulfiller);
+}
+
+TwoPartyVatNetwork::TwoPartyVatNetwork(
+      kj::Own<capnp::MessageStream>&& stream,
+      uint maxFdsPerMessage,
+      rpc::twoparty::Side side,
+      ReaderOptions receiveOptions)
+  : TwoPartyVatNetwork(*stream, kj::mv(stream), maxFdsPerMessage, side, receiveOptions) {}
```

Comment:
I bet that different compilers make difference decisions about whether `*stream` or `kj::mv(stream)` should be evaluated first. If the `Own<>` is moved away too early then `*stream` will dereference a null pointer.

This is... kind of a pain to solve.

I think one hacky solution would be to have yet another intermediate constructor that takes `MessageStream&, kj::Own<MessageStream>&&` (without `Maybe` on the second param yet). This way, `stream` does not actually get moved away on this line; rather, an rvalue reference to `stream` is passed down. The new constructor can, in turn, forward to the constructor that takes `MessageStream&, kj::Maybe<kj::Own<MessageStream>>&&`.

Otherwise, the only alternative may be to live with two copies of the constructor.

Reply:
Pushed a slightly different solution, which goes back to using a kj::OneOf; I think this is easier to follow while still not duplicating code.

Reply:
Hmm, I guess this is bikeshedding at this point but to me having to call `getStream()` which does a switch on every usage seems harder to follow... but ok, it doesn't really matter I guess.